Feat: add support for CSV input data loading from file or inline

Themiscodes opened the pull request:
This is WIP. Feedback is welcome.
- Adds support for CSV external file loading and inline CSV rows, as well as yaml from external file.
- Still to add tests and Docs to be updated accordingly.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/TobikoData/sqlmesh?pullRequest=2640) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/TobikoData/sqlmesh?pullRequest=2640) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/TobikoData/sqlmesh?pullRequest=2640) it.</sub>

georgesittas reviewed (commented):
Cool, looks good. Some general comments to keep in mind w.r.t. code style & conventions. The direction here is good, let's add some tests.

Themiscodes commented on sqlmesh/core/test/definition.py:
```diff
@@ -763,3 +765,32 @@ def _normalize_df_value(value: t.Any) -> t.Any:
+
+def _load_csv(path: Path | str | None, rows: str | None) -> t.List | None:
+    """Handler to load rows from csv file or string."""
+    return pd.read_csv(path or StringIO(rows)).to_dict(orient="records")
```
Got it should I alter it so the `values` are passed in the loaders or leave it for later on when the functionality is expanded to support these `csv_settings`?

georgesittas replied:
It'll be simpler if we [do this inline](https://github.com/TobikoData/sqlmesh/pull/2640#discussion_r1607406946) like Toby suggested, so let's go with that for now.

Themiscodes commented on sqlmesh/core/test/definition.py:
```diff
@@ -763,3 +765,32 @@ def _normalize_df_value(value: t.Any) -> t.Any:
+def _load_rows(values: t.Dict) -> t.List | None:
+    """Load data in rows from external file or inline declared string."""
+    rows = values.get("rows")
+    format = values.get("format") or Path(values.get("path", "")).suffix[1:]
```
Ok, since the inference will be removed for now, when inline rows are provided it will be handled as YAML. 

But when there is a path (without format) should it handle it as YAML and try and load the file or should this not be handled and enforce that the user should have defined the test correctly?

georgesittas replied:
We could try to load the YAML, @tobymao what's your preference here?

georgesittas commented:
@Themiscodes let's make sure to test `NULL` behavior in CSVs as well. In YAML, when you leave out a column from a row, set its value to `null`, or leaving its value unspecified (i.e. `col: `), this results in a `NULL` value when converting it. We should think about how we want to treat empty CSV values.

The problem is that `,,` is ambiguous - it can mean both "empty string" and "null value". Perhaps we want to go with the latter here and if users ever need empty strings, they'd have to use a different format or something.

Themiscodes commented:
> @Themiscodes let's make sure to test `NULL` behavior in CSVs as well. In YAML, when you leave out a column from a row, set its value to `null`, or leaving its value unspecified (i.e. `col: `), this results in a `NULL` value when converting it. We should think about how we want to treat empty CSV values.
> 
> The problem is that `,,` is ambiguous - it can mean both "empty string" and "null value". Perhaps we want to go with the latter here and if users ever need empty strings, they'd have to use a different format or something.

I agree that the latter option seems better. Currently empty columns are handled similar to the yaml and are set to `nan`. If the user wants an empty string it seems more reasonable that they should declare it explicitly.

georgesittas commented on tests/core/test_test.py:
```diff
@@ -501,6 +632,39 @@ def test_partial_data_column_order(sushi_context: Context) -> None:
         ).run()
     )
 
+    _check_successful_or_raise(
```
This doesn't seem particularly useful, CSVs have a fix column order right? Whereas in YAML you can mix it up as shown in the test above.

Themiscodes commented on sqlmesh/core/test/definition.py:
```diff
@@ -328,6 +330,16 @@ def _normalize_rows(
+            elif format == "yaml" or (format is None and path):
+                input_rows = yaml_load(Path(path)) if path else rows
+                rows = input_rows.get("rows") if isinstance(input_rows, dict) else input_rows
+            elif format:
```
Yes you're right since the path check was also after the conditional, changed both of these.

Themiscodes commented:
@georgesittas addressed your comments
1) specifying paths (both csv, yaml) - look into tmp_path etc in other tests 3) supplying a path without a format (expected is to load yaml) 
in: `test_format_path`

2) doing format: yaml and then supplying the rows like we usually do (rows: ...),  
in: `test_format_inline`

4) supplying a format that is not supported
in: `test_unsupported_format_failure`

Also removed unnecessary tests as per your suggestions and parametrized dictionaries where possible.

georgesittas commented on tests/core/test_test.py:
```diff
@@ -390,6 +404,152 @@ def test_partial_data(sushi_context: Context) -> None:
+    )
+
+
+@pytest.mark.parametrize(
+    "input_data",
+    [
+        [
+            """sushi.waiter_names:
+        format: yaml
+        path: """,
+            "test_data.yaml",
+            """- id: 1
+  name: alice
+- id: 2
+  name: 'bob'
+""",
+        ],
+        [
```
You don't have to do this unwrapping from the list manually, `parametrize` offers a cleaner way to achieve this where you can name each part separately:

```python
    ["column", "expected"],
    [
        (  # Array of strings
            "['value1', 'value2']",
            [{"col": ["value1", "value2"]}],
        ),
```

Look into https://github.com/TobikoData/sqlmesh/blob/b11b93e1e19338b007a27c31c0f943327ca08aff/tests/core/test_test.py#L1237

georgesittas reviewed (approved):
Suggestsed a couple of minor refactors, but otherwise LGTM. Thanks for addressing all comments @Themiscodes! 🙌
